### PR TITLE
Move global variables to a ShogunEnv static class

### PR DESCRIPTION
**Pull request by ghost**

### CODE REVIEW

File: examples/meta/generator/targets/cpp.json [hunk 1]
```diff
@@ -119,7 +119,7 @@
             "get_option": "$object->get<std::string>($arguments)",
             "get_string": "$object->get<std::string>($arguments)"
         },
-        "StaticCall": "C$typeName::$method($arguments)",
+        "StaticCall": "$typeName::$method($arguments)",
```
Comment (ghost) [inline examples/meta/generator/targets/cpp.json]: C prefix will be dropped anyways, and there are no static calls now except for ShogunEnv, so I saw no harm

File: src/shogun/io/SGIO.h [hunk 2]
```diff
@@ -544,6 +546,12 @@ class SGIO
 
 	private:
 		RefCount* m_refcount;
+		/// function called to print normal messages
+		std::function<void(FILE*, const char*)> print_message;
+		/// function called to print warning messages
+		std::function<void(FILE*, const char*)> print_warning;
+		/// function called to print error messages
+		std::function<void(FILE*, const char*)> print_error;
```
Comment (ghost) [inline src/shogun/io/SGIO.h:554]: it made more sense that these should be here instead of ShogunEnv

File: src/shogun/lib/Signal.h [hunk 3]
```diff
@@ -30,7 +30,7 @@ namespace shogun
 	 *  option bewteen: immediately exit the running method and fall back to
 	 *  the command line, prematurely stop the current algoritmh and do nothing.
 	 */
-	class CSignal : public CSGObject
+	class CSignal
```
Comment (ghost) [inline src/shogun/lib/Signal.h:33]: doesn't need to inherit from SGObject + it caused problems (ShogunEnv tries to create CSignal, and CSignal tries to create ShogunEnv)

**Comment by vigsterkr** [comment]
i'm a bit ambivalent to make all these functions static functions of the class.... as far as i remember we had problems with one interface implementation calling static functions (maybe R?).

anyhow, the FileSystemRegistry should be as well part of this

File: src/shogun/base/ShogunEnv.h [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+	class Parallel;
+	class SGLinalg;
+	class CSignal;
+
+	class ShogunEnv
+	{
+	public:
+		/** Remove copy and move constructors and assignment operators
+		 */
+		ShogunEnv(const ShogunEnv&) = delete;
```
Comment (vigsterkr) [inline src/shogun/base/ShogunEnv.h]: `SG_DELETE_COPY_AND_ASSIGN` from `shogun/base/macros.h`

File: src/shogun/base/ShogunEnv.h [hunk 5]
```diff
@@ -0,0 +1,131 @@
...
+
+		/** Checks environment variables and modifies global objects
+		 */
+		void init_from_env();
+
+		Parallel* sg_parallel;
+		SGIO* sg_io;
+		Version* sg_version;
+		CSignal* sg_signal;
+		SGLinalg* sg_linalg;
```
Comment (vigsterkr) [inline src/shogun/base/ShogunEnv.h]: why not `std::unique_ptr`?

File: src/shogun/base/ShogunEnv.h [hunk 6]
```diff
@@ -0,0 +1,131 @@
...
+		ShogunEnv();
+
+		/** Checks environment variables and modifies global objects
+		 */
+		void init_from_env();
+
+		Parallel* sg_parallel;
+		SGIO* sg_io;
+		Version* sg_version;
+		CSignal* sg_signal;
```
Comment (vigsterkr) [inline src/shogun/base/ShogunEnv.h]: why not `std::unique_ptr`?

**Comment by karlnapf** [comment]
Yes R and static calls doesn’t work. And it’s a Pitty because since the random stuff is sorted, we don’t need static calls in the examples anymore.....so we could actually enable the R examples now

File: examples/meta/generator/targets/cpp.json (hunk 1 above)
Comment (karlnapf) [reply examples/meta/generator/targets/cpp.json]: I think This should rather be done when the prefix is dropped, not yet

**Comment by ghost** [comment]
The other solution I see is to make only one static function to get the environment, and the other non-static. But this has the same problem with R, and I think it will be more verbose in code (get environment then get what you want).

File: examples/meta/generator/targets/cpp.json (hunk 1 above)
Comment (theartful) [reply examples/meta/generator/targets/cpp.json]: sure, but CShogunEnv is ugly though :"D

**Comment by ghost** [comment]
Does this class need to be exposed at all?

**Comment by vigsterkr** [comment]
imo you dont need to expose this class in the interfaces... or whatever that needs to be used by this class you can wrap those within another class that is exposed to the interface and then call via that... anyhow no need for such things... 
and just to be clear: this class does not need the C prefix as this class is a) not inherited of SGO b) no need to expose it to class_list...

**Comment by ghost** [comment]
This is an attempt using multiple inheritance, please check if you find this approach acceptable @vigsterkr

**Comment by vigsterkr** [comment]
looking nice! :) need to fix some legacy code or?

**Comment by ghost** [comment]
I think this ready. It's worth reminding this it can't be wrapped by swig. I will rebase and push to solve the conflict.

**Comment by vigsterkr** [comment]
@theartful great! good job! as we agree we never wanna expose the env directly to swig... so it should be fine.... and if one day we need to, then we know what to do :)

**Pull request merged**